zcash_client_sqlite/src/wallet.rs

```diff
@@ -0,0 +1,471 @@
+///
+/// This should only be executed inside a transactional context.
+pub fn rewind_to_height<P: consensus::Parameters>(
+    conn: &WalletDB,
+    parameters: &P,
+    block_height: BlockHeight,
+) -> Result<(), SqliteClientError> {
+    let sapling_activation_height = parameters
+        .activation_height(NetworkUpgrade::Sapling)
+        .ok_or(SqliteClientError(Error::SaplingNotActive))?;
+
+    // Recall where we synced up to previously.
+    // If we have never synced, use Sapling activation height.
+    let last_scanned_height =
+        conn.0
+            .query_row("SELECT MAX(height) FROM blocks", NO_PARAMS, |row| {
+                row.get(0)
+                    .map(u32::into)
+                    .or(Ok(sapling_activation_height - 1))
+            })?;
+
+    if block_height >= last_scanned_height {
+        // Nothing to do.
+        return Ok(());
+    }
+
+    // Decrement witnesses.
+    conn.0.execute(
+        "DELETE FROM sapling_witnesses WHERE block > ?",
+        &[u32::from(block_height)],
+    )?;
+
+    // Un-mine transactions.
+    conn.0.execute(
+        "UPDATE transactions SET block = NULL, tx_index = NULL WHERE block > ?",
+        &[u32::from(block_height)],
+    )?;
+
+    // Now that they aren't depended on, delete scanned blocks.
+    conn.0.execute(
+        "DELETE FROM blocks WHERE height > ?",
+        &[u32::from(block_height)],
+    )?;
+
+    Ok(())
+}
+
+pub fn get_commitment_tree(
+    data: &WalletDB,
+    block_height: BlockHeight,
+) -> Result<Option<CommitmentTree<Node>>, SqliteClientError> {
+    data.0
+        .query_row_and_then(
+            "SELECT sapling_tree FROM blocks WHERE height = ?",
+            &[u32::from(block_height)],
+            |row| {
+                let row_data: Vec<u8> = row.get(0)?;
+                CommitmentTree::read(&row_data[..]).map_err(|e| {
+                    rusqlite::Error::FromSqlConversionFailure(
+                        row_data.len(),
+                        rusqlite::types::Type::Blob,
+                        Box::new(e),
+                    )
+                })
+            },
+        )
+        .optional()
+        .map_err(SqliteClientError::from)
+}
+
+pub fn get_witnesses(
+    data: &WalletDB,
+    block_height: BlockHeight,
+) -> Result<Vec<(NoteId, IncrementalWitness<Node>)>, SqliteClientError> {
+    let mut stmt_fetch_witnesses = data
+        .0
+        .prepare("SELECT note, witness FROM sapling_witnesses WHERE block = ?")?;
+    let witnesses = stmt_fetch_witnesses
+        .query_map(&[u32::from(block_height)], |row| {
+            let id_note = NoteId(row.get(0)?);
+            let data: Vec<u8> = row.get(1)?;
+            Ok(IncrementalWitness::read(&data[..]).map(|witness| (id_note, witness)))
+        })
+        .map_err(SqliteClientError::from)?;
+
+    let mut res = vec![];
+    for witness in witnesses {
+        // unwrap database error & IO error from IncrementalWitness::read
+        res.push(witness??);
+    }
```

Comment:
It's more efficient to do this like in the original code (a single allocation instead of something like `log(witnesses.len())`):
```suggestion
    // unwrap database error & IO error from IncrementalWitness::read
    let res: Vec<_> = witnesses.collect::<Result<Result<_, _>, _>>()??;
```

Reply:
I had the damnedest time trying to get `collect` to compile here, but your solution works.